### ISSUE TITLE
Kobo Sage flips multiple pages on single button press

### ISSUE THREAD

**Issue by mkozlows** [koreader/koreader#10743]
* KOReader version: 2023.06.1
* Device: Kobo Sage

#### Issue

Intermittently (<10% of the time, maybe like 5%?), when I push the physical down button to turn a page, the device keeps racing ahead through page after page, until I push the up button to stop it. After doing that, the next press of the down button has no effect, and after that it works normally.

It doesn't seem to be a physically sticky button, because it happens with the reader in either left-handed or right-handed orientation, so it's two separate buttons. (And it doesn't feel stuck down in any way.)

I thought this was fixed at one point, but I was wrong, and it came back. @NiLuJe  mentioned that verbose logs would be interesting to see (https://www.mobileread.com/forums/showpost.php?p=4329019&postcount=453), and so here they are.

(There is a related thing that happens to me sometimes, which is that pushing the down button once will skip through two page turns; this doesn't require me to do anything to stop it, and the button just works normally after it. I've been assuming it's the same issue, but maybe that's not true.)

#### Steps to reproduce
Unfortunately, it can't be reliably reproduced.  But at the very end of the attached crash.log, the following sequence of actions should be visible:

1. I pushed the down button a single time, and a bunch of very fast page turns happened.
2. I pushed the up button a single time to stop the pages from turning.
3. I turned off the screen (I'm pretty sure) and plugged it into a USB port.
4. I turned it back on.
5. I hit the down button a single time to no effect, just to capture it for the log. I might then have hit it again to flip a page forward to get a normal click in there, too, I forget.
6. I went to the menu to turn on USB Mass Storage to get the crash.log.

##### `crash.log` (if applicable)
[crash.log](https://github.com/koreader/koreader/files/12157074/crash.log)

**Comment by NiLuJe** [comment]
> (There is a related thing that happens to me sometimes, which is that pushing the down button once will skip through two page turns; this doesn't require me to do anything to stop it, and the button just works normally after it. I've been assuming it's the same issue, but maybe that's not true.)

The above related to this issue in particular.

I've never experienced nor heard about the other, FWIW. Not home until Friday, so I can't really parse the logs right now.

**Comment by NiLuJe** [comment]
Assuming you're left handed, since the only repeat event I guess is on a page *back* button; and it's an actual hardware repeat event, so it looks like a simple hardware switch snafu to me.

**Comment by mkozlows** [comment]
Yeah, I was holding it in my left hand here. "It looks for all the world like the button is just stuck down" is what the logs looked like to me, and I was hoping there was some subtle intricacy that I was missing, because it's definitely not physically stuck, and it does happen in both orientations (so with both buttons), so if it's a physical problem, it's a weird one somewhere upstream of the physical switch.  

But since nobody else seems to be reporting it, defective unit does seem like the most likely possibility, alas. Thanks for checking.

**Comment by NiLuJe** [comment]
Yep, something like a debouncing issue with the switches or something

**Comment by NiLuJe** [comment]
If you don't actually rely on button repeats anywhere, a simple workaround
would be to patch that out. It *should* be fairly trivial, I'll take a look
when I get back on Friday.

**Comment by NiLuJe** [comment]
You might still get a few chatter of single presses, depending on how
borked the switch is, though.

**Comment by lamidez** [comment]
Par intermittence (<10% du temps, peut-être 5%?), lorsque j’appuie sur le bouton physique vers le bas pour tourner une page, l’appareil continue de courir page après page, jusqu’à ce que j’appuie sur le bouton haut pour l’arrêter. Après cela, l’appui suivant sur le bouton bas n’a aucun effet, et après cela, cela fonctionne normalement.

Moi aussi je rencontre le même problème que Mkozlows, moi uniquement sur les fichiers kepub, avec les fichiers epub je n'ai pas ce problème, moi aussi je souhaite une solution, merci d'avance.

**Comment by NiLuJe** [comment]
@Iamidez: Ça ressemble encore plus à un switch défectueux que ce qui est décrit plus haut; comme toujours, besoin de logs détaillés pour confirmer (mais, encore un fois, jamais vu ça arriver perso, et ça sent vraiment, mais vraiment le matériel foireux).

----

TL;DR: Also looks like a misbehaving switch (i.e., hw issue).

**Comment by mkozlows** [comment]
I will say, after coming to the "probably a hardware defect" conclusion earlier, I was kind of wondering if there was any point to reaching out to Rakuten support (the conclusion appears to be: lol no), but quickly realized that I'd have to demonstrate the problem on the stock software. And so I tried to reproduce it there, and... I can't. It happens in KOReader with the frequency mentioned above, but in the stock Kobo reader, it doesn't occur.

Which makes me wonder if this isn't one of those messy things where there's a tighter tolerance on something within KOReader than in the stock software, such that a switch that's on the outer margin of compliance works fine there, but less so here?

**Comment by NiLuJe** [comment]
I don't recall if Nickel actually handles KEY_REPEAT events, so I'll have to check that ;).

(And... it's a switch, by design, there's no real concept of tolerance on the software side: it's either 0 (open) or 1 (closed)¹, that's the whole point of a switch ;).)

----

[1]: And by KEY_REPEAT (which is actually 2 ;p), what happens is that the kernel (i.e., software) will start emitting those every n ms as long as the switch stays in the "contact/closed" position.

**Comment by NiLuJe** [comment]
> [1]: And by KEY_REPEAT (which is actually 2 ;p), what happens is that the kernel (i.e., software) will start emitting those every n ms as long as the switch stays in the "contact/closed" position.

Specifically, every 80ms after 400ms ;).

```
Key repeat handling:
  Repeat type 20 (EV_REP)
    Repeat code 0 (REP_DELAY)
      Value    400
    Repeat code 1 (REP_PERIOD)
      Value     80
```

----



> I don't recall if Nickel actually handles KEY_REPEAT events, so I'll have to check that ;).

In usual fashion, "it"s weird":
* In RMSDK (i.e., ePubs), repeats are *ignored*
* In ACCESS (i.e., KePubs), repeats are *handled*

In any case, I wanted to post a userpatch to disable repeats for you, but I guess I can just make that a builtin setting, so I'll do that.

**Comment by mkozlows** [comment]
(By "tolerance" I was thinking of having to tweak e.g. debounce values in qmk for keyboard stuff to prevent repeat letters. But I have no idea how applicable that concept is to the particulars of these very different switches.)

As always, appreciate the informative response to half-assed noodling!

**Comment by NiLuJe** [comment]
Oh, yeah, debouncing should *very much* not be our job in userland (I mean, for severely broken stuff on both a hardware and lower-level software, we *could*, but that's terrible. And, yeah, we have a way to do that for single taps in the touch input codepaths because of stuff like that).

In any case, this isn't a debouncing issue, as your switch is not chattering, it's doing the other thing a broken switch can: it sticks closed ;).

**Comment by mkozlows** [comment]
Coming back to this belatedly: The feature you added for disabling repeats solved my "runaway paging" hardware defect problem, really appreciate that. I'm going to close this, with thanks.

(I do still see the "occasionally it flips two pages instead of one" thing, though, which in an earlier comment you had mentioned was a different issue. It sounds like you already  know what's going on there, but let me know if there are any logs that would be useful with that.)

**Comment by NiLuJe** [comment]
Yeah, I have a few ideas to try for the broken refreshes on resume from standby, but haven't had time to really do anything with the Sage yet ;).
